plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceDelay.hs

```diff
@@ -147,20 +154,32 @@ import Data.Foldable as Foldable (foldl')
  detailed above. For implementation details see 'optimisationProcedure'.
 -}
 forceDelay
-    :: Monad m
-    => Term name uni fun a
+    :: (uni ~ DefaultUni, fun ~ DefaultFun, Monad m)
+    => BuiltinSemanticsVariant fun
+    -> Term name uni fun a
     -> SimplifierT name uni fun a m (Term name uni fun a)
-forceDelay term = do
-    let result = transformOf termSubterms processTerm term
+forceDelay semVar term = do
+    let result = transformOf termSubterms (processTerm semVar) term
     recordSimplification term ForceDelay result
     return result
 
 {- | Checks whether the term is of the right form, and "pushes"
  the 'Force' down into the underlying lambda abstractions.
 -}
-processTerm :: Term name uni fun a -> Term name uni fun a
-processTerm = \case
+processTerm
+    :: (uni ~ DefaultUni, fun ~ DefaultFun)
+    => BuiltinSemanticsVariant fun -> Term name uni fun a -> Term name uni fun a
+processTerm semVar = \case
     Force _ (Delay _ t) -> t
+    -- Remove @Delay@s from @ifThenElse@ branches if the latter is @Force@d and the delayed term are
+    -- pure and work-free anyway.
```

Comment:
Should try the same for `chooseList`/`chooseData`/`caseList`/`caseData`, but let's start with the obvious thing and do it only for `ifThenElse` for now.